### PR TITLE
feat(pagination-nav): add component

### LINKED ISSUE #11630: feat(pagination-nav): create component
https://jsw.ibm.com/browse/ADCMS-4835

### Contribution type

Components

### Contribution type (other)

_No response_

### Intent

We'd like to create a web component version of `@carbon/react`'s **Pagination Nav** component.

### Solution

Create a new component called PaginationNav in `@carbon/web-components`. It will have feature parity with the `@carbon/react` version, and it will get its styles from `@carbon/styles`.

Storybook: https://react.carbondesignsystem.com/?path=/story/components-paginationnav--default
Source code: https://github.com/carbon-design-system/carbon/tree/main/packages/react/src/components/PaginationNav
Styles: https://github.com/carbon-design-system/carbon/tree/main/packages/styles/scss/components/pagination-nav

To Do:
- [ ] Create component
- [ ] Create Storybook demos
- [ ] Add e2e tests

### IBM Organization or Company you are part of

Innovation Team, AEM site development team, IBM.com

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/carbon-design-system/carbon-for-ibm-dotcom/blob/main/.github/CODE_OF_CONDUCT.md)
- [X] I checked the [current issues](https://github.com/carbon-design-system/carbon-for-ibm-dotcom/issues) for duplicate issues

### PR DESCRIPTION
**Pull request by andy-blum**
### Related Ticket(s)

Closes #11630 
https://jsw.ibm.com/browse/ADCMS-4835

### Description

Adds `<bx-pagination-nav>` component

### Changelog

**New**

- Adds `<bx-pagination-nav>` component

### CODE REVIEW

File: packages/carbon-web-components/src/components/pagination-nav/pagination-nav.ts [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+   */
+  update(changedProperties) {
+    if (changedProperties.has('count') || changedProperties.has('visible') || changedProperties.has('page')) {
+      const { count, visible: visibleLimit, page } = this;
+      const allItems = [...new Array(Math.max(count, 1))].map((_u, i) => i);
+      let iterator;
+
+      if (allItems.length > visibleLimit) {
+        const first = allItems.shift();
+        const last = allItems.pop();
```
Comment (m4olivei) [inline packages/carbon-web-components/src/components/pagination-nav/pagination-nav.ts:117]: In the React version, first and last can end up being a group. Doesn't seem like that's possible here?

image

**Review by m4olivei — COMMENTED** [review]
Here are some initial notes from a static review.

File: packages/carbon-web-components/src/components/pagination-nav/pagination-nav.ts [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+      _iterator: iterator,
+      setIndex,
+      incrementIndex,
+      decrementIndex,
+    } = this;
+
+    return html`
+      <nav>
+        <ul>
+          <li><button type="button" @click=${decrementIndex}>prev</button></li>
```
[the remark this answers is not in the document]
Comment (andy-blum) [reply packages/carbon-web-components/src/components/pagination-nav/pagination-nav.ts:212]: I'm not worried about the elements/styles used yet - only the functionality.

File: packages/carbon-web-components/src/components/pagination-nav/pagination-nav.ts (hunk 1 above)
Comment (andy-blum) [reply packages/carbon-web-components/src/components/pagination-nav/pagination-nav.ts:117]: I'm pretty sure this is _not_ the intended functionality. @kennylam can you weigh in on this?

Comment (andy-blum) [reply packages/carbon-web-components/src/components/pagination-nav/pagination-nav.ts:117]: Reached out to the carbon-react team for guidance here: https://ibm-marketing-comms.slack.com/archives/C2K6RFJ1G/p1712670269654919

File: packages/carbon-web-components/src/components/pagination-nav/pagination-nav.ts [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+   * Whether to allow looping between first/last pages.
+   */
+  @property({ attribute: "loop", type: Boolean})
+  loop = false;
+
+  /**
+   * Active page index (0-based).
+   */
+  @property({ attribute: "page", reflect: true, type: Number })
+  page = 0;
```
Comment (jkaeser) [inline packages/carbon-web-components/src/components/pagination-nav/pagination-nav.ts:70]: Being 0-based, I can see this confusing users. Thoughts on making it 1-based?

**Review by jkaeser — COMMENTED** [review]
Tested, and I wasn't able to break the component's functionality at all. Feels solid!

File: packages/carbon-web-components/src/components/pagination-nav/pagination-nav.ts (hunk 3 above)
Comment (andy-blum) [reply packages/carbon-web-components/src/components/pagination-nav/pagination-nav.ts:70]: I'm not opposed, but this matches the @carbon/react functionality

Comment (jkaeser) [reply packages/carbon-web-components/src/components/pagination-nav/pagination-nav.ts:70]: Fair enough. +1 for consistency.

File: packages/carbon-web-components/src/components/pagination-nav/pagination-nav.ts (hunk 1 above)
Comment (andy-blum) [reply packages/carbon-web-components/src/components/pagination-nav/pagination-nav.ts:120]: https://github.com/carbon-design-system/carbon/issues/15830#issuecomment-2030066687

**Review by jkaeser — CHANGES_REQUESTED** [review]
@andy-blum This is looking really great! I didn't find any functional issues this time around, either, and I'm only withholding my thumbs up over some minor docs issues.

A few other musings for you...

1. If selecting a page causes the page list to be recalculated, focus doesn't shift to follow the newly active page. What do you think about trying to reset focus on select?
2. Given our recent focus on it, you might consider adding `part`s to the major markup elements.
3. I know you're importing the vast majority of styles, but I wonder if there's anything we can do to help when this appears in small viewports/containers. Maybe we just say this is a content author's responsibility?

https://github.com/carbon-design-system/carbon-for-ibm-dotcom/assets/12532727/03dcbb62-f3bd-45ff-bb60-5fd8455362ef

File: packages/carbon-web-components/src/components/pagination-nav/pagination-nav.ts [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+          newValue: this.page,
+        }
+      })
+      this.dispatchEvent(event);
+      window.requestIdleCallback(()=>{
+        const activeItem = (this.shadowRoot!.querySelector('.bx--pagination-nav__page--active') as HTMLElement);
+        if (activeItem && !activeItem.matches(':focus')) {
+          activeItem.focus();
+        }
+      })
```
Comment (jkaeser) [inline packages/carbon-web-components/src/components/pagination-nav/pagination-nav.ts:139]: Though this has the desired effect when activating an item, it also steals focus when the component is mounted. Also, `requestIdleCallback` isn't yet supported in Safari – https://caniuse.com/requestidlecallback.

**Review by jkaeser — CHANGES_REQUESTED** [review]
Thanks for the updates! Found one more thing that needs addressing.

**Review by jkaeser — APPROVED** [review]

**Review by kennylam — APPROVED** [review]
This is great, thanks @andy-blum!

**Pull request merged**